ImportError

Hi there 👋,

40e8a6fa9db68405cbed31c877d4e92ffe25aaf2 imports type annotations directly from `typing_extensions` instead of doing

```python
import typing
try:
    import typing_extensions
except ImportError:
    pass
from typing import Self
```

[as the documentation suggests](https://typing-extensions.readthedocs.io/en/latest/#runtime-use-of-types).


Note that the typing-extensions dependency is only included for Python < 3.11

https://github.com/confluentinc/confluent-kafka-python/blob/6c3bd6d61fbdaf1afdaaa9de0f48ec45739228d4/requirements/requirements.txt#L1

So if you install the package using Python 3.12, you'll end up with the following ImportError:

```python
    from confluent_kafka.aio import AIOConsumer
.venv/lib/python3.14/site-packages/confluent_kafka/aio/__init__.py:15: in <module>
    from ._AIOConsumer import AIOConsumer
.venv/lib/python3.14/site-packages/confluent_kafka/aio/_AIOConsumer.py:20: in <module>
    from typing_extensions import Self
E   ModuleNotFoundError: No module named 'typing_extensions'
```

[source](https://github.com/codingjoe/django-health-check/actions/runs/24465860001/job/71492597863?pr=716)

Cheers!
Joe

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,7 @@
 - Fix URL joining in Python client (#2228)
 - Handle anyOf/allOf in JSON transforms (#2237)
 - Fix redefining a named Avro type in a diamond dependency pattern (#2238)
+- Fix typing_extensions import errors on Python 3.12 (#2232)
 
 
 ## v2.14.0 - 2026-04-01

--- a/src/confluent_kafka/aio/_AIOConsumer.py
+++ b/src/confluent_kafka/aio/_AIOConsumer.py
@@ -16,8 +16,11 @@ import asyncio
 import concurrent.futures
 from typing import Any, Callable, Dict, Optional, Tuple
 
-# FIXME: import from typing once we depend on Python >= 3.11
-from typing_extensions import Self
+try:
+    from typing import Self
+except ImportError:
+    # FIXME: drop fallback once we require Python >= 3.11
+    from typing_extensions import Self
 
 import confluent_kafka
 

--- a/src/confluent_kafka/aio/_AIOConsumer.py
+++ b/src/confluent_kafka/aio/_AIOConsumer.py
@@ -16,8 +16,11 @@ import asyncio
 import concurrent.futures
 from typing import Any, Callable, Dict, Optional, Tuple
 
-# FIXME: import from typing once we depend on Python >= 3.11
-from typing_extensions import Self
+try:
+    from typing import Self
+except ImportError:
+    # FIXME: import from typing once we depend on Python >= 3.11
+    from typing_extensions import Self
 
 import confluent_kafka
 

--- a/src/confluent_kafka/aio/producer/_AIOProducer.py
+++ b/src/confluent_kafka/aio/producer/_AIOProducer.py
@@ -17,8 +17,11 @@ import concurrent.futures
 import logging
 from typing import Any, Callable, Dict, Optional
 
-# FIXME: import from typing once we depend on Python >= 3.11
-from typing_extensions import Self
+try:
+    from typing import Self
+except ImportError:
+    # FIXME: drop fallback once we require Python >= 3.11
+    from typing_extensions import Self
 
 import confluent_kafka
 

--- a/src/confluent_kafka/aio/producer/_AIOProducer.py
+++ b/src/confluent_kafka/aio/producer/_AIOProducer.py
@@ -17,8 +17,11 @@ import concurrent.futures
 import logging
 from typing import Any, Callable, Dict, Optional
 
-# FIXME: import from typing once we depend on Python >= 3.11
-from typing_extensions import Self
+try:
+    from typing import Self
+except ImportError:
+    # FIXME: import from typing once we depend on Python >= 3.11
+    from typing_extensions import Self
 
 import confluent_kafka
 

--- a/src/confluent_kafka/cimpl.pyi
+++ b/src/confluent_kafka/cimpl.pyi
@@ -35,9 +35,13 @@ maintenance burden and get type hints directly from the implementation.
 """
 
 import builtins
-from typing import Any, Callable, Dict, List, Optional, Tuple, Union, overload
+from typing import Any, Callable, Dict, List, Literal, Optional, Tuple, Union, overload
 
-from typing_extensions import Literal, Self
+try:
+    from typing import Self
+except ImportError:
+    # FIXME: drop fallback once we require Python >= 3.11
+    from typing_extensions import Self
 
 from confluent_kafka.admin._metadata import ClusterMetadata, GroupMetadata
 

--- a/src/confluent_kafka/cimpl.pyi
+++ b/src/confluent_kafka/cimpl.pyi
@@ -35,9 +35,13 @@ maintenance burden and get type hints directly from the implementation.
 """
 
 import builtins
-from typing import Any, Callable, Dict, List, Optional, Tuple, Union, overload
+from typing import Any, Callable, Dict, List, Literal, Optional, Tuple, Union, overload
 
-from typing_extensions import Literal, Self
+try:
+    from typing import Self
+except ImportError:
+    # FIXME: import from typing once we depend on Python >= 3.11
+    from typing_extensions import Self
 
 from confluent_kafka.admin._metadata import ClusterMetadata, GroupMetadata
 

